(WIP) Python3

rr- opened the pull request:
I am porting Trackma to Python 3 (#120) and I came to point where I need guidance before I continue.

What seems to work so far and passed superficial tests:

- MAL
- Curses
- CLI
- QT

What hasn't been tested:

- Anything but MAL

Plus there's a controversy around GTK, which is the reason why I'm presenting this in early and unfinished form.

Seemingly to use GTK in Py3 one needs to use PyGI, as PyGTK for Python 3 never happened. However, there are some rough edges around GTK2 support in PyGI:

- Stocks seem to be unimplemented?
- `Gtk.Menu` has no `.popup` method (but it has it in GTK3)
- `Gdk` has no `SELECTION_CLIPBOARD` which makes us need to pass magic constants to `Gtk.Clipboard.get()`

I'm not sure if we want to cater for GTK2 and start doing weird hacks, or just go along with GTK3.

z411 commented:
Thank you. Seems good.
The thing about GTK seems concerning; I'm not really interested in implementing a GTK3 interface, but if it's the only clean way to do it I guess it's the only way. The other choice is to drop GTK altogether but I'm pretty sure several people use it.

hbers commented:
More than several people in my opinion and personally as a user of the GTK interface, dropping GTK support would be a bad move for this project because most desktop environments use GTK rather than Qt.

rr- commented:
Alright I'll continue work on PyGTK -> PyGI port then. Please note that GTK2 *does* mostly work, it's just cumbersome in certain places.

z411 commented:
Thank you, I think that's safe for now. If it starts getting too cumbersome to maintain, we can eventually work on porting it to GTK3 on a different branch and merge that when it's ready.

Please tell me when this PR is ready to merge. I think I'll bump the version number before merging.

rr- commented:
https://news.ycombinator.com/item?id=12038717 meh

dyskette commented:
@rr- https://wiki.gnome.org/Projects/GTK+/Lifecycle

rr- commented:
Python3's GObject fails to discover `Menu.popup()` method. Curiously, it works without any changes in Trackma if we use GTK3.

rr- commented:
Alright, GTK2 was just too buggy. I changed the GTK interface to use GTK3, which fixes problems with popup menus, clipboard and other things.

I think the PR is ready for alpha testing by some volunteers. I roughly tested:

*Frontend*

- Urwid
- QT
- GTK

*API*

- MAL
- VNDB

What definitely needs to be tested is the rest of the APIs supported by Trackma, plus all possible UI pathways. Unfortunately I don't have any accounts on the remainder of the sites. At the same time I'm somewhat opposed to creating throwaway accounts, especially given that I don't have very high confidence in my tests. So my request is for people to take this PR, test it and report any problems.

dyskette commented:
I'm getting this error, in both qt and gtk:
```
$ trackma-gtk
Data: Locking database...
Data: Reading metadata...
Data: Reading queue...
Data: Processing queue...
Data: Reading cache...
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib64/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 824, in task_start_engine
    self.engine.start()
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 202, in start
    (self.api_info, self.mediainfo) = self.data_handler.start()
  File "/usr/lib/python3.5/site-packages/trackma/data.py", line 153, in start
    self.process_queue()
  File "/usr/lib/python3.5/site-packages/trackma/data.py", line 367, in process_queue
    showid = item['id']
KeyError: 'id'
```

rr- commented:
@dyskette please clear the cache by removing `~/.trackma` (it's a good idea to backup whole `~/.trackma`).

dyskette commented:
@rr- Right, thanks. It was a problem in my end.

Another error when trying to open a hummingbird account:

```
Data: Locking database...
libhb: Logging in...
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib64/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 824, in task_start_engine
    self.engine.start()
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 202, in start
    (self.api_info, self.mediainfo) = self.data_handler.start()
  File "/usr/lib/python3.5/site-packages/trackma/data.py", line 163, in start
    self.download_data()
  File "/usr/lib/python3.5/site-packages/trackma/data.py", line 521, in download_data
    self.showlist = self.api.fetch_list()
  File "/usr/lib/python3.5/site-packages/trackma/lib/libhb.py", line 108, in fetch_list
    self.check_credentials()
  File "/usr/lib/python3.5/site-packages/trackma/lib/libhb.py", line 99, in check_credentials
    response = self._request( "/users/authenticate", post={'username': self.username, 'password': self.password} ).read()
  File "/usr/lib/python3.5/site-packages/trackma/lib/libhb.py", line 82, in _request
    post = urllib.urlencode(post)
AttributeError: module 'urllib' has no attribute 'urlencode'
```

rr- commented:
Fixed & fixed

dyskette commented:
Right.

Hummingbird seems to be OK.

However, opening anilist:

```
libanilist: Requesting access token...
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib64/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 825, in task_start_engine
    self.engine.start()
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 202, in start
    (self.api_info, self.mediainfo) = self.data_handler.start()
  File "/usr/lib/python3.5/site-packages/trackma/data.py", line 163, in start
    self.download_data()
  File "/usr/lib/python3.5/site-packages/trackma/data.py", line 521, in download_data
    self.showlist = self.api.fetch_list()
  File "/usr/lib/python3.5/site-packages/trackma/lib/libanilist.py", line 220, in fetch_list
    self.check_credentials()
  File "/usr/lib/python3.5/site-packages/trackma/lib/libanilist.py", line 212, in check_credentials
    self._request_access_token()
  File "/usr/lib/python3.5/site-packages/trackma/lib/libanilist.py", line 164, in _request_access_token
    data = self._request("POST", "auth/access_token", get=param)
  File "/usr/lib/python3.5/site-packages/trackma/lib/libanilist.py", line 147, in _request
    return json.load(response)
  File "/usr/lib64/python3.5/json/__init__.py", line 268, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/usr/lib64/python3.5/json/__init__.py", line 312, in loads
    s.__class__.__name__))
TypeError: the JSON object must be str, not 'bytes'
```

rr- commented:
Seems to be fixed as well

veluria commented:
Double clicking on a show in the Qt interface:

    Traceback (most recent call last):
      File "/usr/lib/python3.5/site-packages/Trackma-0.4-py3.5.egg/trackma/ui/qtui.py", line 1540, in r_details_loaded
        for line in details['extra']:
    KeyError: 'extra'

Sometimes when clicking on a show in the list:

    Traceback (most recent call last):
      File "/usr/lib/python3.5/site-packages/Trackma-0.4-py3.5.egg/trackma/ui/qtui.py", line 1025, in s_show_selected
        self._select_show(show)
      File "/usr/lib/python3.5/site-packages/Trackma-0.4-py3.5.egg/trackma/ui/qtui.py", line 805, in _select_show
        self.generate_episode_menus(self.menu_play, utils.estimate_aired_episodes(show),show['my_progress'])
      File "/usr/lib/python3.5/site-packages/Trackma-0.4-py3.5.egg/trackma/ui/qtui.py", line 916, in generate_episode_menus
        for ep in range(1, max_eps+1):
    TypeError: 'float' object cannot be interpreted as an integer

rr- commented:
@z411 I'm not sure how to handle issues with incompatible cache (version 2: stores keys as bytes, version 3: stores keys as strings), any ideas? For now I tell the users to remove the cache manually... In the future it might be good to version the cache...

@veluria please try clearing `~/.trackma/*/*.list` cache files. (The `.cache` directory I referred to earlier contains only images and is of no relevance to the error you're experiencing.) Furthermore, could you post titles + site with entries that cause `'float' object cannot be interpreted as integer`? Thanks

veluria commented:
>Furthermore, could you post titles + site with entries that cause 'float' object cannot be interpreted as integer? Thanks

All shows that have no finale episode count (0).

rr- commented:
Fixed the problem with selecting rows

dyskette commented:
GTK, when doing a search in the "Search" window in anilist... but I just could reproduce this 2 times.

```
libanilist: Searching for rep...
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 2401, in select_show
    self.info.load(self.showdict[self.selected_show])
KeyError: 1274
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 2401, in select_show
    self.info.load(self.showdict[self.selected_show])
KeyError: 1274
...
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 2401, in select_show
    self.info.load(self.showdict[self.selected_show])
KeyError: 5763
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 2401, in select_show
    self.info.load(self.showdict[self.selected_show])
KeyError: 5763
```

And this when switching accounts from anilist to hummingbird.
```
libhb: Initializing...
Data: Using libhb (anime)
Data: Locking database...
Data: Reading metadata...
Data: Reading info DB...
Data: Reading cache...
Tracker: Initializing...
Tracker: Enabling tracker...
Tracker: inotifyx not available; using polling (slow).
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 284, in get_show_info
    return showdict[showid]
KeyError: 50

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 893, in select_show
    show = self.engine.get_show_info(self.selected_show)
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 286, in get_show_info
    raise utils.EngineError("Show not found.")
trackma.utils.EngineError: Show not found.
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 284, in get_show_info
    return showdict[showid]
KeyError: 7647

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 893, in select_show
    show = self.engine.get_show_info(self.selected_show)
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 286, in get_show_info
    raise utils.EngineError("Show not found.")
trackma.utils.EngineError: Show not found.
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 284, in get_show_info
    return showdict[showid]
KeyError: 7647

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 893, in select_show
    show = self.engine.get_show_info(self.selected_show)
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 286, in get_show_info
    raise utils.EngineError("Show not found.")
trackma.utils.EngineError: Show not found.
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 284, in get_show_info
    return showdict[showid]
KeyError: 7647

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 893, in select_show
    show = self.engine.get_show_info(self.selected_show)
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 286, in get_show_info
    raise utils.EngineError("Show not found.")
trackma.utils.EngineError: Show not found.

```

Both are just internal and don't crash or stop trackma-gtk.

rr- commented:
Worked around the problem with accessing `showdict` from GTK UI, couldn't reproduce the one during account switch. I think they're more or less harmless, seeing as there were no visible consequences of the last bug I fixed.

@dyskette Regarding closing the application: do you by any chance have "minimize to tray" option selected in settings and use a tiling window manager (or one that doesn't have tray)?

dyskette commented:
@rr- It's actually GNOME Wayland.

I deleted ~/.trackma folder so the "minimize to tray" was a default option that I didn't notice. However, GNOME Wayland does have a tray, and trackma-qt uses it.

--
Error when detecting video playing, trackma-gtk always print "Found player but show not in list". trackma-qt does detect the shows.

```
Tracker: inotifyx not available; using polling (slow).
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/lib64/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.5/site-packages/trackma/tracker.py", line 249, in _tracker
    self._observe_polling(interval)
  File "/usr/lib/python3.5/site-packages/trackma/tracker.py", line 202, in _observe_polling
    self.update_show_if_needed(state, show_tuple)
  File "/usr/lib/python3.5/site-packages/trackma/tracker.py", line 266, in update_show_if_needed
    self._emit_signal('playing', show['id'], True, episode)
  File "/usr/lib/python3.5/site-packages/trackma/tracker.py", line 101, in _emit_signal
    self.signals[signal](*args)
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 132, in _tracker_playing
    self._emit_signal('playing', show, playing, episode)
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 144, in _emit_signal
    self.signals[signal](*args)
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 718, in playing_show
    self.show_lists[status].playing(show, is_playing)
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 1424, in playing
    row[9] = getColor(self.colors['is_playing'])
  File "/usr/lib64/python3.5/site-packages/gi/overrides/Gtk.py", line 1091, in __setitem__
    self.model.set_value(self.iter, key, value)
  File "/usr/lib64/python3.5/site-packages/gi/overrides/Gtk.py", line 991, in set_value
    value = self._convert_value(column, value)
  File "/usr/lib64/python3.5/site-packages/gi/overrides/Gtk.py", line 875, in _convert_value
    return GObject.Value(self.get_column_type(column), value)
  File "/usr/lib64/python3.5/site-packages/gi/overrides/GObject.py", line 218, in __init__
    self.set_value(py_value)
  File "/usr/lib64/python3.5/site-packages/gi/overrides/GObject.py", line 274, in set_value
    (py_value, type(py_value)))
ValueError: Expected string but got Gdk.Color(red=27756, green=11565, blue=51143)<class 'gi.overrides.Gdk.Color'>
```
--
Error when searching show in MAL.

```
libmal: Searching for bliss...
Exception in thread Thread-10:
Traceback (most recent call last):
  File "/usr/lib64/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 2409, in task_search
    self.entries = self.engine.search(self.searchtext.get_text())
  File "/usr/lib/python3.5/site-packages/trackma/engine.py", line 329, in search
    return self.data_handler.search(str(criteria).strip())
  File "/usr/lib/python3.5/site-packages/trackma/data.py", line 202, in search
    results = self.api.search(criteria)
  File "/usr/lib/python3.5/site-packages/trackma/lib/libmal.py", line 229, in search
    root = self._parse_xml(data)
  File "/usr/lib/python3.5/site-packages/trackma/lib/libmal.py", line 702, in _parse_xml
    data = re.sub('<\?.*?\?>', '', data)
  File "/usr/lib64/python3.5/re.py", line 182, in sub
    return _compile(pattern, flags).sub(repl, string, count)
TypeError: cannot use a string pattern on a bytes-like object
```
--
When closing trackma-gtk.

```
(trackma-gtk:22566): Gtk-CRITICAL **: gtk_event_controller_reset: assertion 'GTK_IS_EVENT_CONTROLLER (controller)' failed
```
--
I can't test melative or shikimori, but apart from that I only notice cosmetic issues with GTK3.

Main trackma-gtk window:

[Current status](https://cloud.githubusercontent.com/assets/6687927/17079665/12a4ab0a-50dd-11e6-9afe-845a5e8d9d6b.png)

- Changing progress buttons to a spinbutton.
- Status combobox width smaller.
- Set "Title" column default width, to have a consistent width between tabs.

[So, it looks like this.](https://cloud.githubusercontent.com/assets/6687927/17079666/12a77eac-50dd-11e6-9198-40b9154e1562.png)

"Show Details" window:

[Screenshot](https://cloud.githubusercontent.com/assets/6687927/17079996/b537095e-50e6-11e6-8b6d-758a75ad7917.png)

- Buttons too big.
- Picture too big.

"Global settings" window:

[Screenshot](https://cloud.githubusercontent.com/assets/6687927/17080032/52e7d1ba-50e7-11e6-95be-a737ba2f1f7e.png)

- Height, below buttons "Cancel" and "Apply".

rr- commented:
### General

>Error when detecting video playing, trackma-gtk always print "Found player but show not in list". trackma-qt does detect the shows.

This does not depend on the chosen UI. Please retest with current fixes

>`ValueError: Expected string but got Gdk.Color(red=27756, green=11565, blue=51143)<class 'gi.overrides.Gdk.Color'>`

Fixed

>`TypeError: cannot use a string pattern on a bytes-like object`

Fixed

>`(trackma-gtk:22566): Gtk-CRITICAL **: gtk_event_controller_reset: assertion 'GTK_IS_EVENT_CONTROLLER (controller)' failed`

IDK what's causing it, got it randomly and doesn't seem to do harm

### Main window

>Changing progress buttons to a spinbutton.

The current widget is actually a result of adapting GTK3 (imho it's better - with the GTK2 widget, it was very easy to missclick)

>Status combobox width smaller.

Fixed

>Set "Title" column default width, to have a consistent width between tabs.

Fixed

### Details window

>Buttons too big.

Hopefully fixed, couldn't reproduce

>Picture too big.

Hopefully fixed, couldn't reproduce

### Global settings window

>Height, below buttons "Cancel" and "Apply"

Hopefully fixed, couldn't reproduce

dyskette commented:
GNOME on X.Org, when right click on icon tray.

```
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/trackma/ui/gtkui.py", line 595, in status_menu_event
    menu.popup(None, None, None, Gtk.status_icon_position_menu, button, time, self.statusicon)
  File "/usr/lib64/python3.5/site-packages/gi/overrides/__init__.py", line 39, in __getattr__
    return getattr(self._introspection_module, name)
  File "/usr/lib64/python3.5/site-packages/gi/module.py", line 139, in __getattr__
    self.__name__, name))
AttributeError: 'gi.repository.Gtk' object has no attribute 'status_icon_position_menu'
```

- The buttons to "Accept", "Add" or "Cancel" are misaligned in all windows. Comparison: [GTK2](https://cloud.githubusercontent.com/assets/6687927/17080588/eb2289b2-50fa-11e6-9db5-1a9e182a9350.png) & [GTK3](https://cloud.githubusercontent.com/assets/6687927/17080589/eb262f4a-50fa-11e6-8dd5-349f0dd1dbea.png)

- Details window: Picture too big.

It's actually a hummingbird issue, as it doesn't happens with MAL or anilist.

- Set "Title" column default width.

It's not fixed in my end. Re-tested with clean config and there are hidden columns.

- Changing progress buttons to a spinbutton.

Well, this could be addresed later. My point is that it also fixes these useless warnings: [1](https://cloud.githubusercontent.com/assets/6687927/17080667/01648890-50fd-11e6-8b07-59a953086629.png) & [2](https://cloud.githubusercontent.com/assets/6687927/17080666/016104d6-50fd-11e6-8fe8-b8967dee7ce4.png). But as this is a UI change, it probably should be discussed within its own issue thread instead of a pull request.

z411 commented:
@rr-:
The cache is versioned. Just increment by 1 the `version` variable in the data.py file in [line 40](https://github.com/z411/trackma/blob/7ca267824d9249c57ee8956452723bd1c4fb8c0d/trackma/data.py#L40).
That will make the clients to clear the list cache automatically next start.

Although it won't clear the queue; if you want to change that, you might to change

    if self._queue_exists():

to

    if self._queue_exists() and self.meta.get('version') == self.version:

in [line 136](https://github.com/z411/trackma/blob/7ca267824d9249c57ee8956452723bd1c4fb8c0d/trackma/data.py#L136).

rr- commented:
- Added cache versioning
- Fixed GTK tray icon
- Fixed GTK button alignment
- Fixed GTK picture size being too big
- Fixed clicking on list header to show columns
- Rebased and squashed the commits within the pull request where it made sense

Regarding window title expanding - I don't notice problems unless I start playing with visible columns, but Py2 trackma glitched with these as well.

veluria commented:
After making some changes to the Qt interface config, I get the following message:

>Erroneous config ~/.trackma/ui-qt.json requires manual fixing or deletion to proceed.

rr- commented:
>Erroneous config ~/.trackma/ui-qt.json requires manual fixing or deletion to proceed.

Fixed (issue was with remembering column layout for Qt)

dyskette commented:
I think that's all I can find for now in GTK UI.

In README and docs/gettingstarted.rst:
- Probably should say something about inotifyx not supporting python3.
- PyGTK to PyGI. (python3-gi or python3-gobject)
- Anything python-* to python3-*
- On Fedora PIL is python3-pillow, I think Debian is python3-pil.
- pip3
- python3 setup.py develop

veluria commented:
BBCode and HTML entities are not parsed/displayed in the Gtk interface. No idea whether this was different before.

z411 commented:
inotifyx not working sounds kind of critical, I think we might need to find an alternative. inotifyx seems to be abandoned anyway.

Something like this
https://github.com/dsoprea/PyInotify
I think it supports Python 3.
Or this https://github.com/seb-m/pyinotify

We only need to watch for IN_OPEN and IN_CLOSE events.

z411 commented:
Tested it, seems to be working fine so far, thank you. Is it ready for merging?

dyskette commented:
At least `seb-m/pyinotify` is already packaged in Fedora as [python3-inotify](https://github.com/seb-m/pyinotify) and Debian as [python3-pyinotify](https://packages.debian.org/search?keywords=python3-pyinotify).

rr- commented:
@z411 before merging:

- I'd like to look into the mentioned BBCode
- I'd like to make the suggested adjustments to the docs
- what about `inotifyx` - shouldn't changing it to sth newer go into this pull request?

z411 commented:
@rr-

> I'd like to look into the mentioned BBCode

Indeed that was never implemented; it has nothing to do with the porting to Python 3. So if we are to implement it, opening a new issue/request would be ideal.

Bear in mind that different sites might use BBcode differently, if any, so this might need to be standardized at the lib* level; maybe to HTML. TUIs would need to discard it, though, since they don't use formatting.

> I'd like to make the suggested adjustments to the docs

Oh, please do, although I might be dropping the docs as they're mostly a duplicate of the README and I've abandoned them. Trackma isn't complex enough to require more than a Readme I'd say.

>what about inotifyx - shouldn't changing it to sth newer go into this pull request?

I'd prefer to work on that on a separate issue/request.

I tagged the current master snapshot as v0.5, so with this merge would mark the start of Trackma 0.6.

rr- commented:
Alrighty then, I'll adjust the docs in a few hours, then I'm good to merge. I think a few bugs will still pop out but I'll fire PRs for them as they come (feel free to assign issues / ping me) :3

rr- commented:
Pushed the doc changes to the extent I can verify (I don't use Debian, I used https://www.debian.org/distrib/packages#view to get the package names).
One last thing - I don't think anyone has tested Shikimori and Melative, so I went ahead and created dummy accounts. I fixed one problem with Shikimori and confirmed it works. Then I noticed the same (buggy) code in Melative, which I fixed, but was unable to test it due to not receiving confirmation e-mail.

z411 commented:
I'll be merging this soon; I'll talk to package maintainers first. Thanks.

z411 commented:
Thank you.
